File: kora-app-annotation-processor/src/main/java/io/koraframework/kora/app/annotation/processor/GraphFileGenerator.java
package io.koraframework.kora.app.annotation.processor;

import com.palantir.javapoet.*;
import io.koraframework.annotation.processor.common.AnnotationUtils;
import io.koraframework.annotation.processor.common.CommonClassNames;
import io.koraframework.annotation.processor.common.NameUtils;
import io.koraframework.kora.app.annotation.processor.component.ComponentDependency;
import io.koraframework.kora.app.annotation.processor.component.DependencyClaim;
import io.koraframework.kora.app.annotation.processor.component.ResolvedComponent;
import io.koraframework.kora.app.annotation.processor.declaration.ComponentDeclaration;
import io.koraframework.kora.app.annotation.processor.declaration.ModuleDeclaration;
import io.koraframework.kora.app.annotation.processor.interceptor.ComponentInterceptors;

import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.PackageElement;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.*;
import java.util.function.Supplier;

public class GraphFileGenerator {
    private final ProcessingContext ctx;
    private final Element classElement;
    private final List<TypeElement> allModules;
    private final ComponentInterceptors interceptors;
    private final List<ResolvedComponent> components;
    private final Map<ClassName, ResolvedComponent> conditions;

    public GraphFileGenerator(ProcessingContext ctx, Element classElement, List<TypeElement> allModules, ComponentInterceptors interceptors, List<ResolvedComponent> components, Map<ClassName, ResolvedComponent> conditions) {
        this.ctx = ctx;
        this.classElement = classElement;
        this.allModules = allModules;
        this.interceptors = interceptors;
        this.components = components;
        this.conditions = conditions;
    }

    public JavaFile generate() {
        var packageElement = (PackageElement) classElement.getEnclosingElement();
        var implClass = ClassName.get(packageElement.getQualifiedName().toString(), "$" + classElement.getSimpleName().toString() + "Impl");
        var graphName = classElement.getSimpleName().toString() + "Graph";
        var graphTypeName = ClassName.get(packageElement.getQualifiedName().toString(), graphName);
        var classBuilder = TypeSpec.classBuilder(graphName)
            .addAnnotation(AnnotationUtils.generated(KoraAppProcessor.class))
            .addModifiers(Modifier.PUBLIC)
            .addSuperinterface(ParameterizedTypeName.get(ClassName.get(Supplier.class), CommonClassNames.applicationGraphDraw))
            .addField(CommonClassNames.applicationGraphDraw, "graphDraw", Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
            .addMethod(MethodSpec
                .methodBuilder("get")
                .addAnnotation(Override.class)
                .addModifiers(Modifier.PUBLIC)
                .returns(CommonClassNames.applicationGraphDraw)
                .addStatement("return graphDraw")
                .build());
        var currentClass = (TypeSpec.Builder) null;
        var currentConstructor = (MethodSpec.Builder) null;
        int holders = 0;
        var componentsList = new ArrayList<>(components);
        for (int i = 0; i < componentsList.size(); i++) {
            var componentNumber = i % KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS;
            if (componentNumber == 0) {
                if (currentClass != null) {
                    currentClass.addMethod(currentConstructor.build());
                    classBuilder.addType(currentClass.build());
                    var prevNumber = ((i / KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS) - 1);
                    classBuilder.addField(graphTypeName.nestedClass("ComponentHolder" + prevNumber), "holder" + prevNumber, Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL);
                }
                holders++;
                var className = graphTypeName.nestedClass("ComponentHolder" + i / KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS);
                currentClass = TypeSpec.classBuilder(className)
                    .addAnnotation(AnnotationUtils.generated(KoraAppProcessor.class))
                    .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL);
                currentConstructor = MethodSpec.constructorBuilder()
                    .addModifiers(Modifier.PUBLIC)
                    .addParameter(CommonClassNames.applicationGraphDraw, "graphDraw")
                    .addParameter(implClass, "impl")
                    .addStatement("var map = new $T<$T, $T>()", HashMap.class, String.class, Type.class)
                    .beginControlFlow("for (var field : $T.class.getDeclaredFields())", className)
                    .addStatement("if (!field.getName().startsWith($S)) continue", "component")
                    .addStatement("map.put(field.getName(), (($T) field.getGenericType()).getActualTypeArguments()[0])", ParameterizedType.class)
                    .endControlFlow();
                for (int j = 0; j < i / KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS; j++) {
                    currentConstructor.addParameter(graphTypeName.nestedClass("ComponentHolder" + j), "ComponentHolder" + j);
                }
            }
            var component = componentsList.get(i);
            TypeName componentTypeName = TypeName.get(component.type()).box();
            var typeMirrorElement = ctx.types.asElement(component.type());
            if (typeMirrorElement instanceof TypeElement te) {
                var annotation = AnnotationUtils.findAnnotation(typeMirrorElement, CommonClassNames.aopProxy);
                if (annotation != null) {
                    var superElement = ctx.types.asElement(te.getSuperclass());
                    var aopProxyName = NameUtils.generatedType(superElement, "_AopProxy");
                    if (typeMirrorElement.getSimpleName().contentEquals(aopProxyName)) {
                        componentTypeName = TypeName.get(te.getSuperclass()).box();
                    }
                }
            }

            currentClass.addField(FieldSpec.builder(ParameterizedTypeName.get(CommonClassNames.node, componentTypeName), component.fieldName(), Modifier.PRIVATE, Modifier.FINAL).build());
            currentConstructor.addStatement("var _type_of_$L = map.get($S)", component.fieldName(), component.fieldName());
            var statement = this.generateComponentStatement(graphTypeName, component);
            currentConstructor.addStatement(statement);
        }
        if (components.size() > 0) {
            var lastComponentNumber = components.size() / KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS;
            if (components.size() % KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS == 0) {
                lastComponentNumber--;
            }
            currentClass.addMethod(currentConstructor.build());
            classBuilder.addType(currentClass.build());
            classBuilder.addField(graphTypeName.nestedClass("ComponentHolder" + lastComponentNumber), "holder" + lastComponentNumber, Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL);
        }


        var staticBlock = CodeBlock.builder();

        staticBlock
            .addStatement("var impl = new $T()", implClass)
            .addStatement("graphDraw = new $T($T.class)", CommonClassNames.applicationGraphDraw, classElement);
        for (int i = 0; i < holders; i++) {
            staticBlock.add("$N = new $T(graphDraw, impl", "holder" + i, graphTypeName.nestedClass("ComponentHolder" + i));
            for (int j = 0; j < i; j++) {
                staticBlock.add(", holder" + j);
            }
            staticBlock.add(");\n");
        }

        var supplierMethodBuilder = MethodSpec.methodBuilder("graph")
            .returns(CommonClassNames.applicationGraphDraw)
            .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
            .addStatement("return graphDraw");


        return JavaFile.builder(packageElement.getQualifiedName().toString(), classBuilder
                .addMethod(supplierMethodBuilder.build())
                .addStaticBlock(staticBlock.build())
                .build())
            .build();
    }

    private CodeBlock parentCondition(ResolvedComponent component) {
        if (component.parentConditions().size() == 1) {
            var conditionComponent = Objects.requireNonNull(conditions.get(component.parentConditions().iterator().next()));
            return CodeBlock.of("g.condition($L)", conditionComponent.nodeRef("_"));
        } else {
            var b = CodeBlock.builder();
            b.add("$T.or(", CommonClassNames.graphCondition);
            var parentConditions = new ArrayList<>(component.parentConditions());
            for (int i = 0; i < parentConditions.size(); i++) {
                if (i > 0) {
                    b.add(", ");
                }
                var condition = parentConditions.get(i);
                var conditionComponent = Objects.requireNonNull(conditions.get(condition));
                b.add("g.condition($L)", conditionComponent.nodeRef("_"));
            }
            b.add(")");
            return b.build();
        }
    }

    private CodeBlock generateComponentStatement(ClassName graphTypeName, ResolvedComponent component) {
        var statement = CodeBlock.builder();
        var declaration = component.declaration();
        var componentHolder = component.holderName();
        var componentField = component.fieldName();
        statement.add("$L = graphDraw.addNode(_type_of_$L, ", componentField, componentField);
        if (component.tag() == null) {
            statement.add("null, \n");
        } else {
            statement.add("$L.class, \n", component.tag());
        }

        if (component.parentConditions().isEmpty() && component.declaration().condition() == null) {
            statement.add("null,\n");
        } else if (component.parentConditions().isEmpty()) {
            var conditionComponent = Objects.requireNonNull(conditions.get(component.declaration().condition()));
            statement.add("g -> g.condition($L).eval(),\n", conditionComponent.nodeRef("_"));
        } else if (component.declaration().condition() == null) {
            statement.add("g -> $L.eval(),\n", parentCondition(component));
        } else {
            var conditionComponent = Objects.requireNonNull(conditions.get(component.declaration().condition()));
            statement.add("g -> $T.and($L, g.condition($L)).eval(),\n", CommonClassNames.graphCondition, parentCondition(component), conditionComponent.nodeRef("_"));
        }

        var createDependencies = this.getCreateDependencies(componentHolder, component);
        statement.add("$L,\n", createDependencies);

        var refreshDependencies = this.getRefreshDependencies(componentHolder, component);
        statement.add("$L,\n", refreshDependencies);

        var interceptorsFor = interceptors.interceptorsFor(component);
        statement.add("$T.of(", List.class);
        for (int i = 0; i < interceptorsFor.size(); i++) {
            if (i > 0) {
                statement.add(", ");
            }
            var interceptor = interceptorsFor.get(i);
            statement.add("$L", interceptor.component().nodeRef(componentHolder));
        }
        statement.add("),\n");

        statement.add("g -> ");
        var dependenciesCode = this.generateDependenciesCode(ctx, component, graphTypeName);

        switch (declaration) {
            case ComponentDeclaration.AnnotatedComponent annotatedComponent -> {
                statement.add("new $T", ClassName.get(annotatedComponent.typeElement()));
                if (!annotatedComponent.typeVariables().isEmpty()) {
                    statement.add("<");
                    for (int i = 0; i < annotatedComponent.typeVariables().size(); i++) {
                        if (i > 0) statement.add(", ");
                        statement.add("$T", annotatedComponent.typeVariables().get(i));
                    }
                    statement.add(">");
                }
                statement.add("($L)", dependenciesCode);
            }
            case ComponentDeclaration.FromModuleComponent moduleComponent -> {
                if (moduleComponent.module() instanceof ModuleDeclaration.AnnotatedModule(var element)) {
                    statement.add("impl.module$L.", allModules.indexOf(element));
                } else {
                    statement.add("impl.");
                }
                if (!moduleComponent.typeVariables().isEmpty()) {
                    statement.add("<");
                    for (int i = 0; i < moduleComponent.typeVariables().size(); i++) {
                        if (i > 0) statement.add(", ");
                        statement.add("$T", moduleComponent.typeVariables().get(i));
                    }
                    statement.add(">");
                }
                statement.add("$L($L)", moduleComponent.method().getSimpleName(), dependenciesCode);
            }
            case ComponentDeclaration.FromExtensionComponent extension -> statement.add(extension.generator().apply(dependenciesCode));
            case ComponentDeclaration.PromisedProxyComponent promisedProxyComponent -> {
                if (promisedProxyComponent.typeElement().getTypeParameters().isEmpty()) {
                    statement.add("new $T($L)", promisedProxyComponent.className(), dependenciesCode);
                } else {
                    statement.add("new $T<>($L)", promisedProxyComponent.className(), dependenciesCode);
                }
            }
            case ComponentDeclaration.OptionalComponent optional -> {
                var optionalOf = ((DeclaredType) optional.type()).getTypeArguments().get(0);
                statement.add("$T.<$T>ofNullable($L)", Optional.class, optionalOf, dependenciesCode);
            }
            case null, default -> throw new RuntimeException("Unknown type " + declaration);
        }
        statement.add(")");
        return statement.build();

    }


    private CodeBlock generateDependenciesCode(ProcessingContext ctx, ResolvedComponent component, ClassName graphTypeName) {
        var resolvedDependencies = component.dependencies();
        if (resolvedDependencies.isEmpty()) {
            return CodeBlock.of("");
        }
        var b = CodeBlock.builder();
        b.indent();
        b.add("\n");
        for (int i = 0, dependenciesSize = resolvedDependencies.size(); i < dependenciesSize; i++) {
            if (i > 0) b.add(",\n");
            var resolvedDependency = resolvedDependencies.get(i);
            b.add(resolvedDependency.write(ctx, graphTypeName));
        }
        b.unindent();
        b.add("\n");
        return b.build();
    }

    private CodeBlock getCreateDependencies(String componentHolder, ResolvedComponent component) {
        var result = new ArrayList<ResolvedComponent>();
        if (component.declaration().condition() != null) {
            var condition = Objects.requireNonNull(this.conditions.get(component.declaration().condition()));
            result.add(condition);
        }
        for (var parentConditionTag : component.parentConditions()) {
            var condition = Objects.requireNonNull(this.conditions.get(parentConditionTag));
            result.add(condition);
        }
        for (var dependency : component.dependencies()) {
            switch (dependency) {
                case ComponentDependency.NullDependency _, ComponentDependency.PromisedProxyParameterDependency _ -> {}
                case ComponentDependency.SingleDependency singleDependency -> {
                    switch (singleDependency) {
                        case ComponentDependency.PromiseOfDependency _ -> {}
                        case ComponentDependency.TargetDependency targetDependency when targetDependency.claim().claimType() != DependencyClaim.DependencyClaimType.NODE_OF ->
                            result.add(targetDependency.component());
                        case ComponentDependency.TargetDependency _ -> {}
                        case ComponentDependency.ValueOfDependency valueOfDependency -> result.add(valueOfDependency.component());
                        case ComponentDependency.WrappedTargetDependency wrappedTargetDependency -> result.add(wrappedTargetDependency.component());
                    }
                }
                case ComponentDependency.AllOfDependency allOfDependency -> {
                    if (allOfDependency.claim().claimType() != DependencyClaim.DependencyClaimType.ALL_OF_PROMISE) {
                        for (var d : allOfDependency.getResolvedDependencies()) {
                            result.add(d.component());
                        }
                    }
                }
                case ComponentDependency.OneOfDependency oneOfDependency -> {
                    for (var singleDependency : oneOfDependency.dependencies()) {
                        switch (singleDependency) {
                            case ComponentDependency.PromiseOfDependency _ -> {}
                            case ComponentDependency.TargetDependency targetDependency -> result.add(targetDependency.component());
                            case ComponentDependency.ValueOfDependency valueOfDependency -> result.add(valueOfDependency.component());
                            case ComponentDependency.WrappedTargetDependency wrappedTargetDependency -> result.add(wrappedTargetDependency.component());
                        }
                    }
                }
                case ComponentDependency.GraphDependency _, ComponentDependency.TypeOfDependency _ -> {}
            }
        }
        var b = CodeBlock.builder();
        b.add("$T.of(", List.class);
        for (int i = 0; i < result.size(); i++) {
            if (i > 0) {
                b.add(", ");
            }
            b.add("$L", CodeBlock.of("$L", result.get(i).nodeRef(componentHolder)));
        }
        b.add(")");
        return b.build();
    }

    private CodeBlock getRefreshDependencies(String componentHolder, ResolvedComponent component) {
        var result = new ArrayList<ResolvedComponent>();
        if (component.declaration().condition() != null) {
            var condition = Objects.requireNonNull(this.conditions.get(component.declaration().condition()));
            result.add(condition);
        }
        for (var parentConditionTag : component.parentConditions()) {
            var condition = Objects.requireNonNull(this.conditions.get(parentConditionTag));
            result.add(condition);
        }
        for (var dependency : component.dependencies()) {
            switch (dependency) {
                case ComponentDependency.NullDependency _, ComponentDependency.PromisedProxyParameterDependency _ -> {}
                case ComponentDependency.SingleDependency singleDependency -> {
                    switch (singleDependency) {
                        case ComponentDependency.PromiseOfDependency _, ComponentDependency.ValueOfDependency _ -> {}
                        case ComponentDependency.TargetDependency targetDependency when targetDependency.claim().claimType() != DependencyClaim.DependencyClaimType.NODE_OF ->
                            result.add(targetDependency.component());
                        case ComponentDependency.TargetDependency _ -> {}
                        case ComponentDependency.WrappedTargetDependency wrappedTargetDependency -> result.add(wrappedTargetDependency.component());
                    }
                }
                case ComponentDependency.AllOfDependency allOfDependency -> {
                    if (allOfDependency.claim().claimType() == DependencyClaim.DependencyClaimType.ALL_OF_ONE) {
                        for (var resolved : allOfDependency.getResolvedDependencies()) {
                            result.add(Objects.requireNonNull(resolved.component()));
                        }
                    }
                }
                case ComponentDependency.OneOfDependency oneOfDependency -> {
                    for (var singleDependency : oneOfDependency.dependencies()) {
                        switch (singleDependency) {
                            case ComponentDependency.PromiseOfDependency _ -> {}
                            case ComponentDependency.TargetDependency targetDependency -> result.add(targetDependency.component());
                            case ComponentDependency.ValueOfDependency valueOfDependency -> result.add(valueOfDependency.component());
                            case ComponentDependency.WrappedTargetDependency wrappedTargetDependency -> result.add(wrappedTargetDependency.component());
                        }
                    }
                }
                case ComponentDependency.GraphDependency _, ComponentDependency.TypeOfDependency _ -> {}
            }
        }
        var b = CodeBlock.builder();
        b.add("$T.of(", List.class);
        for (int i = 0; i < result.size(); i++) {
            if (i > 0) {
                b.add(", ");
            }
            b.add("$L", result.get(i).nodeRef(componentHolder));
        }
        b.add(")");
        return b.build();
    }

}

File: kora-app-symbol-processor/src/main/kotlin/io/koraframework/kora/app/ksp/GraphFileGenerator.kt
package io.koraframework.kora.app.ksp

import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import io.koraframework.kora.app.ksp.component.ComponentDependency
import io.koraframework.kora.app.ksp.component.DependencyClaim
import io.koraframework.kora.app.ksp.component.ResolvedComponent
import io.koraframework.kora.app.ksp.declaration.ComponentDeclaration
import io.koraframework.kora.app.ksp.declaration.ModuleDeclaration
import io.koraframework.kora.app.ksp.interceptor.ComponentInterceptors
import io.koraframework.ksp.common.CommonClassNames
import io.koraframework.ksp.common.KotlinPoetUtils.controlFlow
import io.koraframework.ksp.common.KspCommonUtils.generated
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type
import java.util.*
import java.util.function.Supplier


class GraphFileGenerator(
    val ctx: ProcessingContext,
    val declaration: KSClassDeclaration,
    val allModules: List<KSClassDeclaration>,
    val interceptors: ComponentInterceptors,
    val components: List<ResolvedComponent>,
    val conditions: Map<ClassName, ResolvedComponent>
) {

    fun generate(): FileSpec {
        val packageName = declaration.packageName.asString()
        val graphName = "${declaration.simpleName.asString()}Graph"
        val graphTypeName = ClassName(packageName, graphName)

        val fileSpec = FileSpec.builder(
            packageName = packageName,
            fileName = graphName
        )

        val implClass = ClassName(packageName, "\$${declaration.simpleName.asString()}Impl")
        val supplierSuperInterface = Supplier::class.asClassName().parameterizedBy(CommonClassNames.applicationGraphDraw)
        val classBuilder = TypeSpec.classBuilder(graphName)
            .generated(KoraAppProcessor::class)
            .addSuperinterface(supplierSuperInterface)
            .addFunction(
                FunSpec.builder("get")
                    .addModifiers(KModifier.OVERRIDE)
                    .returns(CommonClassNames.applicationGraphDraw)
                    .addStatement("return graphDraw")
                    .build()
            )

        val companion = TypeSpec.companionObjectBuilder()
            .generated(KoraAppProcessor::class)
            .addProperty("graphDraw", CommonClassNames.applicationGraphDraw)

        var currentClass: TypeSpec.Builder? = null
        var currentConstructor: FunSpec.Builder? = null
        var holders = 0

        for ((i, component) in components.withIndex()) {
            val componentNumber = i % KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS
            if (componentNumber == 0) {
                if (currentClass != null) {
                    currentClass.primaryConstructor(currentConstructor!!.build())
                    classBuilder.addType(currentClass.build())
                    val prevNumber = i / KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS - 1
                    companion.addProperty("holder$prevNumber", graphTypeName.nestedClass("ComponentHolder$prevNumber"))
                }
                holders++
                val className = graphTypeName.nestedClass("ComponentHolder" + i / KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS)
                currentClass = TypeSpec.classBuilder(className)
                    .generated(KoraAppProcessor::class)
                currentConstructor = FunSpec.constructorBuilder()
                    .addParameter("graphDraw", CommonClassNames.applicationGraphDraw)
                    .addParameter("impl", implClass)
                    .addStatement("val self = %T", graphTypeName)
                    .addStatement("val map = %T<%T, %T>()", HashMap::class.asClassName(), String::class.asClassName(), Type::class.asClassName())
                    .controlFlow("for (field in %T::class.java.declaredFields)", className) {
                        controlFlow("if (!field.name.startsWith(%S))", "component") { addStatement("continue") }
                        addStatement("map[field.name] = (field.genericType as %T).actualTypeArguments[0]", ParameterizedType::class.asClassName())
                    }
                for (j in 0 until i / KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS) {
                    currentConstructor.addParameter("ComponentHolder$j", graphTypeName.nestedClass("ComponentHolder$j"))
                }
            }

            val aopProxySuperClass = ServiceTypesHelper.findAopProxySuperClass(component.type)
            val propertyType = aopProxySuperClass?.toTypeName() ?: component.type.toTypeName()

            currentClass!!.addProperty(component.fieldName, CommonClassNames.node.parameterizedBy(propertyType))
            val statement = this.generateComponentStatement(component)
            currentConstructor!!.addCode(statement).addCode("\n")
        }
        if (components.size > 0) {
            var lastComponentNumber = components.size / KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS
            if (components.size % KoraAppProcessor.COMPONENTS_PER_HOLDER_CLASS == 0) {
                lastComponentNumber--
            }
            currentClass!!.addFunction(currentConstructor!!.build())
            classBuilder.addType(currentClass.build())
            companion.addProperty("holder$lastComponentNumber", graphTypeName.nestedClass("ComponentHolder$lastComponentNumber"))
        }


        val initBlock = CodeBlock.builder()
            .addStatement("val self = %T", graphTypeName)
            .addStatement("val impl = %T()", implClass)
            .addStatement("graphDraw =  %T(%T::class.java)", CommonClassNames.applicationGraphDraw, declaration.toClassName())
        for (i in 0 until holders) {
            initBlock.add("%N = %T(graphDraw, impl", "holder$i", graphTypeName.nestedClass("ComponentHolder$i"))
            for (j in 0 until i) {
                initBlock.add(", holder$j")
            }
            initBlock.add(")\n")
        }

        val supplierMethodBuilder = FunSpec.builder("graph")
            .returns(CommonClassNames.applicationGraphDraw)
            .addCode("\nreturn graphDraw\n", declaration.simpleName.asString() + "Graph")
        return fileSpec.addType(
            classBuilder
                .addType(companion.addInitializerBlock(initBlock.build()).addFunction(supplierMethodBuilder.build()).build())
                .addFunction(supplierMethodBuilder.build())
                .build()
        ).build()
    }


    private fun parentCondition(component: ResolvedComponent): CodeBlock {
        if (component.getParentConditions().size == 1) {
            val conditionComponent = conditions[component.getParentConditions().first()]!!
            return CodeBlock.of("it.condition(%L)", conditionComponent.nodeRef("_"))
        } else {
            val b = CodeBlock.builder()
            b.add("%T.or(", CommonClassNames.graphCondition)
            for ((i, condition) in component.getParentConditions().withIndex()) {
                if (i > 0) {
                    b.add(", ")
                }
                val conditionComponent = conditions[condition]!!
                b.add("it.condition(%L)", conditionComponent.nodeRef("_"))
            }
            b.add(")")
            return b.build()
        }
    }

    private fun generateComponentStatement(component: ResolvedComponent): CodeBlock {
        val statement = CodeBlock.builder()
        val declaration = component.declaration
        val componentHolder = component.holderName
        val componentField = component.fieldName

        statement.add("%N = graphDraw.addNode(map[%S], ", componentField, component.fieldName)
        statement.indent().add("\n")
        if (component.tag == null) {
            statement.add("null,\n")
        } else {
            statement.add("%L::class.java,\n", component.tag)
        }

        if (component.getParentConditions().isEmpty() && component.declaration.condition == null) {
            statement.add("null,\n");
        } else if (component.getParentConditions().isEmpty()) {
            val conditionComponent = conditions[component.declaration.condition]!!
            statement.add("{ it.condition(%L).eval() },\n", conditionComponent.nodeRef("_"));
        } else if (component.declaration.condition == null) {
            statement.add("{ %L.eval() },\n", parentCondition(component));
        } else {
            val conditionComponent = conditions[component.declaration.condition]!!
            statement.add("{ %T.and(%L, it.condition(%L)).eval() },\n", CommonClassNames.graphCondition, parentCondition(component), conditionComponent.nodeRef("_"));
        }


        val createDependencies = getCreateDependencies(componentHolder, component)
        statement.add("%L,\n", createDependencies)

        val refreshDependencies = getRefreshDependencies(componentHolder, component)
        statement.add("%L,\n", refreshDependencies)

        statement.add("listOf(")
        for ((i, interceptor) in interceptors.interceptorsFor(declaration).withIndex()) {
            if (i > 0) {
                statement.add(", ")
            }
            statement.add("%L", interceptor.component.nodeRef(componentHolder))
        }
        statement.add("),\n")


        statement.add("{ ")
        val dependenciesCode = this.getDependenciesCode(component)

        when (declaration) {
            is ComponentDeclaration.AnnotatedComponent -> {
                statement.add("%T", declaration.classDeclaration.toClassName())
                if (declaration.typeVariables.isNotEmpty()) {
                    statement.add("<")
                    for ((i, tv) in declaration.typeVariables.withIndex()) {
                        if (i > 0) {
                            statement.add(", ")
                        }
                        statement.add("%L", tv.type!!.toTypeName())
                    }
                    statement.add(">")
                }
                statement.add("(%L)", dependenciesCode)
            }

            is ComponentDeclaration.FromModuleComponent -> {
                if (declaration.module is ModuleDeclaration.AnnotatedModule) {
                    statement.add("impl.module%L.", allModules.indexOf(declaration.module.element))
                } else {
                    statement.add("impl.")
                }
                statement.add("%N", declaration.method.simpleName.asString())
                if (declaration.typeVariables.isNotEmpty()) {
                    statement.add("<")
                    for ((i, tv) in declaration.typeVariables.withIndex()) {
                        if (i > 0) {
                            statement.add(", ")
                        }
                        statement.add("%L", tv.type!!.toTypeName())
                    }
                    statement.add(">")
                }
                statement.add("(%L)", dependenciesCode)
            }

            is ComponentDeclaration.FromExtensionComponent -> {
                statement.add(declaration.generator(dependenciesCode))
            }

            is ComponentDeclaration.PromisedProxyComponent -> {
                statement.add("%T(%L)", declaration.className, dependenciesCode)
            }

            is ComponentDeclaration.OptionalComponent -> {
                statement.add("%T.ofNullable(%L)", Optional::class.asClassName(), dependenciesCode)
            }
        }
        statement.add(" }\n")
        return statement.unindent().add(")\n").build()
    }

    private fun getCreateDependencies(componentHolder: String, component: ResolvedComponent): CodeBlock {
        val result = mutableListOf<ResolvedComponent>()
        if (component.declaration.condition != null) {
            val condition = this.conditions[component.declaration.condition]!!
            result.add(condition);
        }
        for (parentConditionTag in component.getParentConditions()) {
            val condition = this.conditions[parentConditionTag]!!
            result.add(condition);
        }

        for (dependency in component.dependencies) {
            when (dependency) {
                is ComponentDependency.NullDependency, is ComponentDependency.PromisedProxyParameterDependency -> {}
                is ComponentDependency.TypeOfDependency -> {}
                is ComponentDependency.SingleDependency -> {
                    when (dependency) {
                        is ComponentDependency.PromiseOfDependency -> {}
                        is ComponentDependency.TargetDependency -> result.add(dependency.component)
                        is ComponentDependency.ValueOfDependency -> result.add(dependency.component)

                        is ComponentDependency.WrappedTargetDependency -> result.add(dependency.component)
                    }
                }

                is ComponentDependency.AllOfDependency -> {
                    if (dependency.claim.claimType !== DependencyClaim.DependencyClaimType.ALL_OF_PROMISE) {
                        for (resolvedComponent in dependency.resolvedDependencies) {
                            resolvedComponent.component?.let { result.add(it) }
                        }
                    }
                }

                is ComponentDependency.OneOfDependency -> {
                    for (singleDependency in dependency.dependencies) {
                        when (singleDependency) {
                            is ComponentDependency.PromiseOfDependency -> {}
                            is ComponentDependency.TargetDependency -> result.add(singleDependency.component)
                            is ComponentDependency.ValueOfDependency -> result.add(singleDependency.component)
                            is ComponentDependency.WrappedTargetDependency -> result.add(singleDependency.component)
                            is ComponentDependency.NullDependency -> throw IllegalStateException()
                        }
                    }

                }
            }
        }
        val b = CodeBlock.builder()
        b.add("%M(", MemberName("kotlin.collections", "listOf"))
        for (i in result.indices) {
            if (i > 0) {
                b.add(", ")
            }
            b.add("%L", result[i].nodeRef(componentHolder))
        }
        b.add(")")
        return b.build()
    }


    private fun getRefreshDependencies(
        componentHolder: String,
        component: ResolvedComponent
    ): CodeBlock {
        val result = mutableListOf<ResolvedComponent>()
        if (component.declaration.condition != null) {
            val condition = this.conditions[component.declaration.condition]!!
            result.add(condition);
        }
        for (parentConditionTag in component.getParentConditions()) {
            val condition = this.conditions[parentConditionTag]!!
            result.add(condition);
        }

        for (dependency in component.dependencies) {
            when (dependency) {
                is ComponentDependency.NullDependency, is ComponentDependency.PromisedProxyParameterDependency, is ComponentDependency.TypeOfDependency -> {}
                is ComponentDependency.SingleDependency -> {
                    when (dependency) {
                        is ComponentDependency.TargetDependency -> result.add(dependency.component)
                        is ComponentDependency.WrappedTargetDependency -> result.add(dependency.component)
                        is ComponentDependency.PromiseOfDependency, is ComponentDependency.ValueOfDependency -> {}
                    }
                }

                is ComponentDependency.AllOfDependency -> {
                    if (dependency.claim.claimType === DependencyClaim.DependencyClaimType.ALL) {
                        for (dependencyComponent in dependency.resolvedDependencies) {
                            dependencyComponent.component?.let { result.add(it) }
                        }
                    }
                }

                is ComponentDependency.OneOfDependency -> {
                    for (singleDependency in dependency.dependencies) {
                        when (singleDependency) {
                            is ComponentDependency.PromiseOfDependency -> {}
                            is ComponentDependency.TargetDependency -> result.add(singleDependency.component)
                            is ComponentDependency.ValueOfDependency -> result.add(singleDependency.component)
                            is ComponentDependency.WrappedTargetDependency -> result.add(singleDependency.component)
                            else -> throw IllegalStateException()
                        }
                    }
                }

            }
        }
        val b = CodeBlock.builder()
        b.add("%M(", MemberName("kotlin.collections", "listOf"))
        for (i in result.indices) {
            if (i > 0) {
                b.add(", ")
            }
            b.add("%L", result[i].nodeRef(componentHolder))
        }
        b.add(")")
        return b.build()
    }


    private fun getDependenciesCode(component: ResolvedComponent): CodeBlock {
        if (component.dependencies.isEmpty()) {
            return CodeBlock.of("")
        }
        val block = CodeBlock.builder().indent().add("\n")
        for ((i, dependency) in component.dependencies.withIndex()) {
            if (i > 0) {
                block.add(",\n")
            }
            block.add(dependency.write(ctx))
        }
        block.unindent().add("\n")
        return block.build()
    }

}

